### PR TITLE
feat: an initial client implementation of GA4GH ServiceInfo and TES APIs as a library

### PR DESCRIPTION
**Pull request by aaravm**
An initial client implementation of GA4GH ServiceInfo and TES APIs as a library. 

New structs implementing the current [architecture](https://www.figma.com/board/pSdOr8mnzly8QFOPtnmLhY/ga4gh-sdk-architecture?node-id=0-1&t=j8LIGTmadWSCPZVc-0):

- Configuration: basic struct for storing credentials and client service configuration;
- Transport: HTTP client wrapper that uses Configuration;
- ServiceInfo: implements GA4GH ServiceInfo retrieval; 
- Task: Details of the Task, has functions for cancelling and getting status;
- TES: For creating new tasks, getting a list of them and per-task info;

New features:

-  test covers all the mentioned structs; using Funnel.
-  CI/CD: formatting checks, linting, tests against Funnel,  
-  Automated script for models auto-generation using OpenAPI specs for providing all the definitions of the various structs being used in the above-mentioned structs.

Areas of improvement:

- Passing parameters to Transport methods

## Summary by Sourcery

This pull request introduces an initial client implementation of the GA4GH ServiceInfo and TES APIs as a library. It includes new structs for configuration, transport, service information retrieval, and task management. Additionally, it sets up CI workflows, automated model generation, and comprehensive tests.

- **New Features**:
    - Implemented initial client library for GA4GH ServiceInfo and TES APIs, including structs for Configuration, Transport, ServiceInfo, Task, and TES.
    - Added functionality for creating, listing, and managing tasks within the TES API.
    - Introduced automated script for model auto-generation using OpenAPI specifications.
- **Enhancements**:
    - Wrapped HTTP client with a Transport struct to handle requests using the provided configuration.
- **Build**:
    - Added build script (build-models.sh) for generating models from OpenAPI specifications.
- **CI**:
    - Set up CI workflow to include formatting checks, linting, and tests against Funnel.
- **Documentation**:
    - Added README.md with instructions for building the project, running tests, and using the CI/CD workflow.
- **Tests**:
    - Added comprehensive tests for the new structs and functionalities, including task creation, status retrieval, cancellation, and listing.

### CODE REVIEW

File: lib/src/tes/mod.rs
Comment (pavelnikonorov) [inline lib/src/tes/mod.rs:215]: I believe we have to find a way to implement Transport methods (e.g., `get`) in a way that they can accept any struct as a URL parameter for the provided end-point, like this:

```
    pub async fn list_tasks(
        &self,
        params: Option<ListTasksParams>,
    ) -> Result<TesListTasksResponse, Box<dyn std::error::Error>> {
           self.transport.get("/tasks", params).await;
```
           
The postprocessing of `params` should be done inside Transport, including removing the Null parameters as they cause errors. 

Otherwise, there will be a need to do such complex and custom workarounds whenever we pass parameters to the Trasnport methods

**Review by pavelnikonorov — CHANGES_REQUESTED** [review]
### Request Changes

Thank you for the updates! I've left 11 comments that need to be addressed before we can proceed with merging. Please address the following issues. Here are the key points:

1. **Cargo.toml**: package name and description.
2. **lib/sample**: Move to the upstream folder.
3. **lib/src/tes/mod.rs: ListTasksParams struct**: should be autogenerated along with all other models. 
4. **TES.check() and its usage**: possible improvement for brevity and code reusability.
5.  **Passing parameters to Transport methods**: this is the most essential suggestion that will impact all further development. 
6. **lib/src/tes/mod.rs create_task()**: possibly unnecessary and unused feature.
7. **lib/src/tes/mod.rs test_cancel_task() & test_list_task()**: possibly unnecessary cloning of task.id.

For more detailed feedback, see the inline comments.

Thanks!

File: .github/workflows/ci.yml
Comment (uniqueg) [reply .github/workflows/ci.yml:104]: Also, what about unit tests and code coverage? I think for a project of this scope, we absolutely need to make sure that all code is tested well, both via unit and integration/end-to-end tests.

File: run-tests.sh
Comment (uniqueg) [inline run-tests.sh:14]: Please remove and readd if needed, here and below

File: build-models.sh
Comment (uniqueg) [inline build-models.sh:20]: Please remove and readd if needed (here and below)

File: lib/sample/grape.tes
Comment (uniqueg) [reply lib/sample/grape.tes:1]: I would rather put it in a subfolder of `tests/` in the root directory, e.g., `tests/files/`. I think `tests/` is more standard. Or otherwise `examples/`.

Agreeing on documenting the examples in `README.md`.

Comment (uniqueg) [reply lib/sample/grape.tes:1]: Oh, and I would remove the placeholders that aren't yet supported or are difficult to set up. We can successively build up a more complex example as we go along.

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:9]: Better to create a GitHub issue. Tracking `TODO`s is a pain in collaborative repos.

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:12]: For consistency with `ApiKey`, we could be more explicit and use a `struct` here with members `username` and `password` (or `user` and `pass`).

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:20]: Don't really know Rust, but is it normal to give a "class" (impl) and "model" (struct) the same name?

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:24]: Why can't we set the other members of the struct, like `basic_auth` when creating a new config instance?

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:40]: Shouldn't we have setters for all members?

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:42]: I don't really understand this. Seems to me that we have a constructor that requires 3 args, and a default configuration. But what if I want to create an instance and provide only some args. Can't I set defaults for every value that wasn't provided as an arg instead?

File: lib/src/configuration.rs
Comment (uniqueg) [inline lib/src/configuration.rs:1]: Is it possible that this is just sandboxing here to get something up and running? Not sure if I should review...

File: lib/src/serviceinfo/mod.rs
Comment (uniqueg) [inline lib/src/serviceinfo/mod.rs:1]: Is this ready for review? :thinking:

File: lib/src/tes/mod.rs
Comment (uniqueg) [reply lib/src/tes/mod.rs:19]: Question for my understanding: Shouldn't we commit the autogenerated models to version control? Or are you planning to always generate them on the fly? If the latter, what if we need to manually edit them (and we _will_ need to manually edit them!)

File: lib/src/tes/mod.rs
Comment (uniqueg) [inline lib/src/tes/mod.rs:259]: Okay, seems I don't understand what `grape.tes` is for. Because I am surprised to find that in the models...?

File: lib/src/tes/mod.rs
Comment (uniqueg) [inline lib/src/tes/mod.rs:1]: I think this does way more than just define and validate the TES models, no? If so, I strongly think we need more separation here. Models in one place, controller code and tests elsewhere.

File: lib/src/test_utils.rs
Comment (uniqueg) [inline lib/src/test_utils.rs:32]: I think there should be no mention of Funnel in `lib`. I am not really comfortable with any tight coupling of code to specific implementations. It's (maybe) okay to use it for testing (even though it's not a reference implementation, and so tests against Funnel don't necessarily tell us much, one way or the other), but I think it that should be done strictly _outside_ of the main codebase.

**Review by uniqueg — CHANGES_REQUESTED** [review]
I think it would be great to create a few PRs out of this one, I would recommend the following:
- `chore: add .gitignore`
- `chore: add license` (I recommend Apache Version 2)
- `docs: add initial README.md`
- `ci: add initial GH Actions workflow`
- `build: add Rust build system and deps` (i.e., `Cargo.toml` and `rustfmt.toml`)
- `feat: add OAI model builder`
- `feat: add TES API models`
- `feat: add ServiceInfo API models`
- one or more `feat` PRs to add additional code

I am also missing documentation in the code. Something to generate API docs from. It's very hard to read the code otherwise. Besides, we can use it to define doctests, as I understand.

But maybe I'm just missing something. It's the first time I'm reading Rust code in depth - so apologies if not all of my comments make sense :pray:

File: Cargo.toml
Comment (aaravm) [reply Cargo.toml:6]: Ok, will change it

File: Cargo.toml
Comment (aaravm) [reply Cargo.toml:2]: Ok, changed

File: Cargo.toml
Comment (aaravm) [reply Cargo.toml:30]: Since, library target names cannot contain hyphens, I will put ga4gh_sdk here instead

File: lib/src/tes/mod.rs
Comment (aaravm) [reply lib/src/tes/mod.rs:188]: Yeah, I agree. Do you think we should create an issue and leave it for now, or try resolving it now?

File: build-models.sh
Comment (aaravm) [reply build-models.sh:38]: I am not sure what you mean by this. I think [https://openapi-generator.tech/docs/generators/rust](https://openapi-generator.tech/docs/generators/rust) itself is being used, as specified by -g rust in the command.

File: lib/sample/grape.tes
Comment (pavelnikonorov) [reply lib/sample/grape.tes:1]: Agree @aaravm lets put some generic Task definition file instead of this, from Funnel examples for instance

File: Cargo.toml
Comment (pavelnikonorov) [reply Cargo.toml]: ga4gh_sdk at least should work

File: .github/workflows/ci.yml
Comment (pavelnikonorov) [reply .github/workflows/ci.yml]: yes, 24.04 is available in GitHub CI/CD, @aaravm let's try

https://github.com/actions/runner-images/issues/9848

File: lib/src/configuration.rs
Comment (pavelnikonorov) [reply lib/src/configuration.rs:20]: @uniqueg Yes, this is exactly the way of doing this – we define struct first, which is about data structure fundamentally, then we implement ("impl") methods for this

File: .github/workflows/ci.yml
Comment (pavelnikonorov) [reply .github/workflows/ci.yml]: @uniqueg I think installing Rust is legacy already, as it was added to upgrade the Rust version to enable working with `rustfmt.toml` to exclude autogenerated models from formatting checks, yet it hasn't worked at the end; @aaravm let's try to remove Rust install after switching to ubuntu 24, and `rustfmt.toml` as well if it's not used actually

File: lib/src/test_utils.rs
Comment (aaravm) [reply lib/src/test_utils.rs:32]: I have created a issue regarding this point at https://github.com/elixir-cloud-aai/ga4gh-sdk/issues/15. We can further do the necessary changes in the near future

File: build-models.sh
Comment (pavelnikonorov) [reply build-models.sh:38]: @uniqueg the link you provided describes Rust generator functionality, the tool itself is written in Java. Here it is executed through NPX, the NodeJS Package eXecutor. @aaravm let's switch to a more direct way of using this https://openapi-generator.tech/docs/installation#bash-launcher-script – this executes jar file directly

File: lib/src/configuration.rs
Comment (aaravm) [reply lib/src/configuration.rs:20]: Yeah, exactly. I took reference from here: https://users.rust-lang.org/t/confusion-struct-impl-self-trait/3941/4

File: build-models.sh
Comment (pavelnikonorov) [reply build-models.sh:1]: @uniqueg This script responsibility is much like `configure.sh` in C projects. It might be redundant to implement it in Rust. It has to run an external Java tool, through `npx` (as currently done) or bash script provided by its maintainers (see comment above).  Also, this code is not needed to be in the same code space within the lib, as it's more service-level for the project repo/codebase itself

File: lib/src/tes/mod.rs
Comment (aaravm) [reply lib/src/tes/mod.rs]: This grape.tes is in the tests module only, and is _only being used for tesing_. Currently, this is being used to make a sample task. It is in a function, since this file is being used in multiple tests, so this way redundancy can be avoided. 
To be noted, that this now replaced by sample.tes.

File: lib/src/tes/mod.rs
Comment (aaravm) [reply lib/src/tes/mod.rs:1]: I have created the issue https://github.com/elixir-cloud-aai/ga4gh-sdk/issues/15 to add the tests in different folders, and use other implementations.

Also, currently too the tests module is entirely separate, and even if it is moved to a new file, I don't think any code would need to be changed

File: lib/src/tes/mod.rs
Comment (pavelnikonorov) [reply lib/src/tes/mod.rs]: @uniqueg `grape.tes` is in the `sample` dir, not models, it contains TES Task definition JSON data, which was quickly taken from the [pre-initial CLI implementation](https://github.com/genxnetwork/ga4gh-cli). as agreed in another conversation here, [will be updated to a more generic one and moved](https://github.com/elixir-cloud-aai/ga4gh-sdk/pull/13#discussion_r1665281867) to ./tests/files

File: lib/src/tes/mod.rs
Comment (pavelnikonorov) [reply lib/src/tes/mod.rs:1]: I agree, the `ListTaskParams` certently shouldn't be here. It wans't autogenerated so it has to be manually coded. Probably OpenAPI specs have to updated; Aarav have created an issue for this https://github.com/elixir-cloud-aai/ga4gh-sdk/issues/14 

@aaravm as a quick workaround for this, maybe let's move this model to `model.rs` in the same folder? not in `./models` as `./build-models.sh` fully rewrite this folder

File: lib/sample/grape.tes
Comment (aaravm) [reply lib/sample/grape.tes:1]: I have changed the sample file, and taken a new one from funnel docs: https://github.com/elixir-cloud-aai/ga4gh-sdk/blob/dev/tests/sample.tes

File: lib/src/tes/mod.rs
Comment (pavelnikonorov) [reply lib/src/tes/mod.rs:1]: @uniqueg As for tests, it's common and encouraged in Rust to have **unit** tests near the code they are testing, which can include testing of private functions as well.

But our tests are both unit and integration effectively as they use external systems like Funnel. Should be redesigned then...

```
project
├── Cargo.lock
├── Cargo.toml
├── src
│   └── lib.rs
└── tests
    └── integration_test.rs
```

```
// lib.rs
pub fn add_two(a: i32) -> i32 {
    internal_adder(a, 2)
}

fn internal_adder(a: i32, b: i32) -> i32 {
    a + b
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn internal() {
        assert_eq!(4, internal_adder(2, 2));
    }
}
```

[Rust Doc: Tests](https://doc.rust-lang.org/book/ch11-03-test-organization.html)

**Comment by pavelnikonorov** [comment]
@uniqueg We have discussed your suggestion to split this into multiple PRs with Aarav. Here’s the plan:

1. Address all the suggestions and requested changes, and resolve conversations here.
2. Create a set of branches from the current one, each with relevant features/files.
3. Raise new PRs for each branch.
4. Close this PR with a comment containing the list of new PRs.

File: .github/workflows/ci.yml
Comment (aaravm) [reply .github/workflows/ci.yml]: Sure, ok. Here is a list of the versions I have changed:
actions/cache: v2 -> v3
actions/setup-java: v2 -> v3
actions/setup-go: v2 -> v4
actions/checkout: v2 -> v4

File: lib/src/configuration.rs
Comment (aaravm) [reply lib/src/configuration.rs:24]: I have added basic_auth in the constructor. @pavelnikonorov do you think we should add any other parameter to constructor?

File: lib/src/serviceinfo/mod.rs
Comment (aaravm) [reply lib/src/serviceinfo/mod.rs:1]: Well, I don't see any reason why not

File: lib/src/tes/mod.rs
Comment (aaravm) [reply lib/src/tes/mod.rs]: I think we can keep post processing of params as it is, as earlier I was entering the params in a wrong manner, which caused such long and complex workarounds. I have changed it now and it now takes only 1 line instead of checking NULL values for each attribute in the param, using the function:
```
fn serialize_to_json<T: Serialize>(item: T) -> Value {
    serde_json::to_value(&item).unwrap()
}
```
and adding
```
#[serde(skip_serializing_if = "Option::is_none")]
```
to the ListTasksParams struct

File: .github/workflows/ci.yml
Comment (aaravm) [reply .github/workflows/ci.yml]: I have added an option to run tests using cargo-nextest

Comment (aaravm) [reply .github/workflows/ci.yml]: I have also added unit test coverage using llvm-cov, since can be used with nextest

**Review by pavelnikonorov — APPROVED** [review]
All suggestions addressed, the code is split now into multiple PRs. This PR have to be closed without merge.

File: README.md
Comment (pavelnikonorov) [reply README.md:1]: stubs added in the docs/readme branch

**Comment by aaravm** [comment]
Since, this PR has been successfully split into multiple PRs, I am closing this PR now

**Pull request closed without merging**
